### PR TITLE
tooling for sourcemaps

### PR DESCRIPTION
**Pull request by EduardoRFS**
## Context

As discussed in private I'm working on sourcemaps, but that requires some big changes, so here I added a couple miscellaneous changes that will be used in the near future for sourcemaps.

### locations on J.expression

This just adds Locations in the expressions so that it can be used on Js_dump.

### static dispatch on Ext_pp

This just made it easier to reason for the next changes.

### add Ext_string.fold_left

Helper to find newlines and keep track of that during printing.

### expose current line and column on ext_pp

Track newlines to know which column and line Ext_pp currently is, this is needed as sourcemaps require current position of the printer.

### fuse scope and pp under cxt on js_dump

This is very useful as when adding sourcemaps we would need to pass even another element around, or declare a global variable, both which are terrible ideas. So here I just pass a context record around.

### CODE REVIEW

File: jscomp/ext/ext_pp.ml [hunk 1]
```diff
@@ -29,66 +29,98 @@ end
...
+}
+
+let output_string t s =
+  (match t.kind with
+  | Channel chan -> output_string chan s
+  | Buffer buf -> Buffer.add_string buf s);
+  let (new_line, new_column) =
+    Ext_string.fold_left 
+      (fun (line, column) char ->
+          if char = '\n' then
```
Comment (anmonteiro) [inline jscomp/ext/ext_pp.ml:51]: Are there other line endings we should support? e.g. `\r`

Comment (EduardoRFS) [reply jscomp/ext/ext_pp.ml]: \r is not a valid line ending outside of the original Mac OS, it is always \r\n or just \n, which this handles both

File: jscomp/ext/ext_pp.ml [hunk 2]
```diff
@@ -29,66 +29,88 @@ end
...
   mutable last_new_line : bool;
       (* only when we print newline, we print the indent *)
 }
 
+let output_string t s =
+  (match t.kind with
+  | Channel chan -> output_string chan s
+  | Buffer buf -> Buffer.add_string buf s);
+  let new_line, new_column =
+    Ext_string.fold_left
```
Comment (anmonteiro) [inline jscomp/ext/ext_pp.ml:48]: why is this needed?

Comment (anmonteiro) [reply jscomp/ext/ext_pp.ml:48]: another option is to `String.split` and check `List.length`

**Comment by anmonteiro** [comment]
OK let's do this.
